elastic/kibana
# [Infrastructure UI] Improve Hosts View page performance

## Summary

This PR focuses on enhancing the performance of the Hosts View page to improve the user experience. The main performance detractors were identified as multiple unnecessary calls to the data view's toSpec() function during chart rendering and excessive re-rendering of charts whenever there were changes on the page.

To address these issues, the following optimizations were implemented:

### Caching the toSpec() Result

Previously, the toSpec() function was called multiple times for each chart, resulting in heavy computation. To optimize this,  caching mechanism was introduced that stores the result of the toSpec() function. This prevents unnecessary and repetitive computations, significantly improving the chart attributes generation performance.

_Before_
<img width="900" alt="image" src="https://github.com/elastic/kibana/assets/2767137/c9ed1a0e-c99b-4a09-a091-562739f69abb">

_After_
<img width="900" alt="image" src="https://github.com/elastic/kibana/assets/2767137/7d390f22-2696-48d5-bb41-df849abd6f1b">

To optimize the chart attributes generation, the `toSpec()` result is now cached. Preventing this heavy operation from performing unnecessarily

### Reducing Lens Re-renders

In the previous implementation, metric charts and KPIs were re-rendered excessively whenever there were changes on the page, impacting performance. By implementing memoization techniques, unnecessary re-renders were solved, resulting in smoother and faster rendering of metric charts and KPIs.

**_Metric charts_**

_Before_
<img width="900" alt="image" src="https://github.com/elastic/kibana/assets/2767137/fdb3ac05-1c50-4b2d-bde6-68a9d1601444">


_After_
<img width="900" alt="image" src="https://github.com/elastic/kibana/assets/2767137/c2c5ade9-96df-4c41-be2a-d5323e8fc44d">


**_KPIs_**

_Before_
<img width="900" alt="image" src="https://github.com/elastic/kibana/assets/2767137/75abe5ae-7c2e-4cee-b816-07f31e542d7f">

_After_
<img width="900" alt="image" src="https://github.com/elastic/kibana/assets/2767137/bdf32738-b11d-4da8-b0d7-6792de9ed0a7">


### Overall performance

By implementing the above optimizations, the overall performance of the Hosts View page has been greatly improved. The changes have significantly reduced delays and improved the responsiveness of buttons, checkboxes, and tab interactions.

_Before_
<img width="900" alt="image" src="https://github.com/elastic/kibana/assets/2767137/c390cdc8-4421-46ee-9ea0-dc28ef1efd97">

_After_
<img width="900" alt="image" src="https://github.com/elastic/kibana/assets/2767137/771ee718-fd58-49e0-a377-ce1e187cec3a">


### How to test

- Start a local Kibana instance
- Navigate to `Infrastructure > Hosts`
- Click through buttons, checkboxes tabs and check if there are delays for the click actions to be performed
